drivers/counter/counter_ll_stm32_timer.c

```diff
@@ -0,0 +1,632 @@
+void tim_irq_handler(const struct device *dev)
+{
+	TIM_TypeDef *timer = get_stm32_config(dev)->timer;
+
+	/* Capture compare events */
+	switch (counter_get_num_of_channels(dev)) {
+	case 4:
+		TIM_IRQ_HANDLE_CC(timer, 4);
+	case 3:
+		TIM_IRQ_HANDLE_CC(timer, 3);
+	case 2:
+		TIM_IRQ_HANDLE_CC(timer, 2);
+	case 1:
+		TIM_IRQ_HANDLE_CC(timer, 1);
+	}
+
+	/* TIM Update event */
+	if (LL_TIM_IsActiveFlag_UPDATE(timer) && LL_TIM_IsEnabledIT_UPDATE(timer)) {
+		LL_TIM_ClearFlag_UPDATE(timer);
+		top_irq_handle(dev);
+	}
+}
+
+#define TIMER(idx)              DT_PARENT(DT_DRV_INST(idx))
+
+#define TIMER_IRQ_CONNECT(idx)				 \
+	IRQ_DIRECT_CONNECT(DT_IRQN(TIMER(idx)),		 \
+			   DT_IRQ(TIMER(idx), priority), \
+			   counter##idx##_irq_handler,	 \
+			   IRQ_ZERO_LATENCY)
+
+/** TIMx instance from DT */
+#define TIM(idx) ((TIM_TypeDef *)DT_REG_ADDR(DT_PARENT(DT_DRV_INST(idx))))
+
+#define DT_INST_CLK(idx)				\
+	{						\
+		.bus = DT_CLOCKS_CELL(TIMER(idx), bus),	\
+		.enr = DT_CLOCKS_CELL(TIMER(idx), bits)	\
+	}
+
+#define COUNTER_DEVICE_INIT(idx)						\
+	BUILD_ASSERT(DT_INST_PROP(idx, st_prescaler) <=				\
+		     0xFFFF,							\
+		     "TIMER counter prescaler out of range");			\
+	BUILD_ASSERT(NUM_CH(TIM(idx)) <= TIMER_MAX_CH,				\
+		     "TIMER too many channels");				\
+	static struct counter_stm32_data counter##idx##_data;			\
+	static struct counter_stm32_ch_data					\
+	counter##idx##_ch_data[TIMER_MAX_CH];					\
+	ISR_DIRECT_DECLARE(counter##idx##_irq_handler)				\
+	{									\
+		tim_irq_handler(DEVICE_DT_GET(DT_DRV_INST(idx)));		\
+		return 0;							\
+	}									\
+	static int counter##idx##_init(const struct device *dev)		\
+	{									\
+		TIMER_IRQ_CONNECT(idx);						\
+		return init_timer(dev);						\
+	}									\
+	LOG_INSTANCE_REGISTER(LOG_MODULE_NAME, idx, CONFIG_COUNTER_LOG_LEVEL);	\
+	static struct counter_stm32_config counter##idx##_config = {		\
+		.info = {							\
+			.max_top_value =					\
+				IS_TIM_32B_COUNTER_INSTANCE(TIM(idx)) ?		\
+				0xFFFFFFFF : 0x0000FFFF,			\
+			.freq = 1U,						\
+			.flags = COUNTER_CONFIG_INFO_COUNT_UP,			\
+			.channels = NUM_CH(TIM(idx)),				\
+		},								\
+		.ch_data = counter##idx##_ch_data,				\
+		.timer = TIM(idx),						\
+		.prescaler = DT_INST_PROP(idx, st_prescaler),			\
+		.pclken = DT_INST_CLK(idx),					\
+		.irqn = DT_IRQN(TIMER(idx)),					\
+		LOG_INSTANCE_PTR_INIT(log, LOG_MODULE_NAME, idx)		\
+	};									\
+	DEVICE_DT_INST_DEFINE(idx,						\
+			      counter##idx##_init,				\
+			      NULL,						\
+			      &counter##idx##_data,				\
+			      &counter##idx##_config.info,			\
+			      PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
+			      &counter_stm32_driver_api);
+
+BUILD_ASSERT(!DT_NODE_HAS_STATUS(DT_CHILD(DT_NODELABEL(timers1), counter), okay),
+	     "Advanced-control timer TIM1 unsupported by this driver");
+BUILD_ASSERT(!DT_NODE_HAS_STATUS(DT_CHILD(DT_NODELABEL(timers8), counter), okay),
+	     "Advanced-control timer TIM8 unsupported by this driver");
```

Comment:
Last minute remark:
Why Advanced-control timer are not supported ?
Note: as implementation is no more specific to STM32F4, then TIM20 is also an Advanced-control timer on some other STM32 series.

Reply:
When I enable interrupts for either of the advanced-control timers, it seems to break interrupts for some of the general-purpose ones. E.g. enabling TIM1 will stop interrupts for TIM9. This seems to be hinted at by how the macros are defined, e.g. `TIM1_BRK_TIM9_IRQn` and `TIM8_BRK_TIM12_IRQn` would seem to indicate some overlap with these other timers. So I'm not sure if it's possible to use both at the same time.

Reply:
Ok, it might "just" be because you need to add some IRQ sharing between those peripherals.
This is doable, but could be left for a future PR, as you're achieving already a great work with current one.

Reply:
In this case I propose at least to change the error message: remove the mention to "Advanced-control" as it is not linked to advanced timer (on some STM32 series, like F0, G0, H7, .. , TIM1 IRQ is not multiplexed with other Timer IRQ) and give the exact reason.
For example: "TIM1 unsupported by this driver to avoid potential unimplemented shared IRQ conflict"

Reply:
The error messages have been removed completely, per @erwango's suggestion to rely on "st,stm32-counter" in the device tree instead (so TIM1 and TIM8 are simply not marked as compatible).